File: Difficulty/Medium/[187] Repeated DNA Sequences/solution.cpp
/*
 * @lc app=leetcode id=187 lang=cpp
 *
 * [187] Repeated DNA Sequences
 */

// @lc code=start
class Solution {
public:
    vector<string> findRepeatedDnaSequences(string s) {
        if (s.size() < 10) return vector<string>();
        vector<string> res;
        unordered_map<string, int> um;
        for (int i = 0; i <= s.size() - 10; i++) {
            um[s.substr(i, 10)]++;
        }
        for(auto& p : um) {
            if (p.second > 1)
                res.push_back(p.first);
        }
        return res;
    }
};
// @lc code=end

File: Difficulty/Medium/[187] Repeated DNA Sequences/solution.py
class Solution:
    def findRepeatedDnaSequences(self, s: str) -> List[str]:
        if len(s) < 10:
            return []
        res, d = [], dict()
        for i in range(len(s) - 9):
            d[s[i:i+10]] = d.get(s[i:i+10], 0) + 1
        for k, v in d.items():
            if v > 1:
                res.append(k)
        return res